### ISSUE TITLE
Add Apple Core AI export support

### ISSUE THREAD

**Issue by glenn-jocher** [ultralytics/ultralytics#25620]
## Summary

Track initial Ultralytics export support for Apple's new Core AI `.aimodel` format through [`coreai-torch`](https://github.com/apple/coreai-torch).

A local feasibility spike on 2026-08-02 shows that fixed-shape YOLO26 export is already close, but accelerated runtime validation is blocked until macOS 27 / iOS 27 and Xcode 27 are available. Core ML should remain the supported Apple deployment path until that validation is complete.

## Environment

- Ultralytics `main` at `4e21eac2`
- `coreai-torch==0.4.1`, `coreai-core==1.0.0b2`
- Python 3.12.12, PyTorch 2.11.0
- Apple M5 Pro, macOS 26.6, Xcode 26.6
- Core AI specialization unavailable: `SpecializationOptions.is_supported() == False`

## Conversion results

Each model was fused, exported at its production input size with `torch.export.export()`, decomposed with `coreai_torch.get_decomp_table()`, converted with `TorchConverter`, optimized, and saved with `AIProgram.save_asset()`.

| Model | Input | Outputs | `.aimodel` | Result |
| --- | ---: | --- | ---: | --- |
| `yolo26n.pt` | 640 | `(1, 300, 6)` | 9.860 MB | Success |
| `yolo26n-seg.pt` | 640 | `(1, 300, 38)`, `(1, 32, 160, 160)` | 11.134 MB | Success |
| `yolo26n-sem.pt` | 640 | `(1, 19, 640, 640)` | 6.284 MB | Success |
| `yolo26n-depth.pt` | 640 | `(1, 1, 640, 640)` | 20.770 MB | Success |
| `yolo26n-cls.pt` | 224 | `(1, 1000)` | 11.278 MB | Success |
| `yolo26n-pose.pt` | 640 | `(1, 300, 57)` | 11.942 MB | Success |
| `yolo26n-obb.pt` | 640 | `(1, 300, 7)` | 10.033 MB | Success |

All seven decomposed `ExportedProgram` outputs matched eager PyTorch exactly before Core AI conversion.

The unspecialized Core AI Python runtime also produced numerically close results for all tasks. For pose on `bus.jpg`, the maximum absolute difference across all 300 output rows was `0.002243`; the four detections above 0.25 confidence were within `0.000153`, including keypoints.

## Required graph change

YOLO26 detection-derived heads currently produce `aten.remainder.Scalar` from `index % nc`. `coreai-torch==0.4.1` rejects that overload even though it supports `aten.remainder.Tensor` and `aten.fmod.Scalar`.

A Core AI-specific `torch.fmod(index, nc)` path converts successfully because the top-k indices are nonnegative and `nc` is positive. This must remain specific to `format="coreai"`: replacing the expression globally breaks the existing Core ML exporter because `coremltools==9.0` does not lower `fmod`.

## Benchmark status

An accelerated Core AI vs Core ML comparison is not possible on the available OS/toolchain:

- Core AI model loading works on macOS 26.6, but specialization is unsupported. YOLO26n detect therefore runs through an unspecialized fallback at about **1,584 ms median** (20 timed runs after five warmups). This is a converter/runtime correctness check, not a production performance result.
- The current published Core ML baseline on a physical iPhone 17 Pro is **3.2 ms CPU + ANE preferred** and **9.2 ms CPU-only** for YOLO26n detect. These numbers are not directly comparable to the unspecialized Core AI fallback.
- Local Core ML packages exported with coremltools 9.0 aborted in MPSGraph runtime compilation on this M5 Pro/macOS 26.6 host, including with coremltools' exact tested PyTorch 2.7.0, so no trustworthy same-host Core ML latency is reported.
- The connected iPhone 17 Pro is currently offline and runs iOS 26.6, which cannot exercise the iOS 27 Core AI runtime.

## Proposed initial scope

After macOS 27 / Xcode 27 or an iOS 27 device is available:

- Add `format=coreai` as an export-only, fixed-shape YOLO26 target first.
- Require Python >=3.11 and PyTorch >=2.8 only when this format is selected.
- Start with FP32 and `batch=1`; reject dynamic shapes, embedded NMS, and quantization until each is validated.
- Preserve existing output contracts and names for all seven canonical tasks.
- Add Core AI-specific metadata and a minimal export smoke path.
- Validate numerical parity and real CPU/GPU/ANE placement, latency, memory, first-run specialization time, and cache behavior.
- Add the corresponding loader/preprocessing/output-decoding path to the iOS SDK before describing the format as deployable.

## Recommendation

Do not ship the exporter yet. The conversion surface is promising and broad, but production support needs accelerated macOS 27/iOS 27 validation and an application runtime. Keep this issue as the implementation and benchmark tracker and retain Core ML as the supported Apple format meanwhile.

Related: #25129

**Comment by UltralyticsAssistant** [comment]
👋 Hello @glenn-jocher, thank you for the detailed Core AI export investigation and implementation tracker 🚀! This automated response confirms that your issue has been received, and an Ultralytics engineer will also assist soon.

For general usage and export guidance, we recommend visiting the [Docs](https://docs.ultralytics.com/), including the [Python](https://docs.ultralytics.com/usage/python) and [CLI](https://docs.ultralytics.com/usage/cli) usage examples. You can also review the latest [YOLO26 documentation](https://docs.ultralytics.com/models/yolo26/) while Core ML remains the supported Apple deployment path.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example) to help us debug it.

If this is a custom training ❓ Question, please provide as much information as possible, including dataset image examples and training logs, and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips).

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo26"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes) and [Tasks](https://docs.ultralytics.com/tasks) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

**Comment by john-rocky** [comment]
I have YOLO26 exporting to Apple's Core AI format, with on-device numbers.

The only unsupported op after `run_decompositions()` is `aten.remainder.Scalar`, from `(index % nc)` in the end2end head's postprocess. A four-line decomposition into integer floor-division handles it exactly, no fmod approximation needed:

```python
def remainder_scalar(x, s):
    if x.dtype in (torch.int64, torch.int32, torch.int16, torch.int8):
        return x - s * torch.div(x, s, rounding_mode="floor")
    return x - s * torch.floor(x / s)

table = coreai_torch.get_decomp_table()
table[torch.ops.aten.remainder.Scalar] = remainder_scalar
```

All five task heads then convert and match PyTorch on a real image at 640 (coreai-torch 0.4.2): detect 9.9MB, segment 11.1MB, pose 12.0MB, obb 10.0MB, classify 11.3MB, worst matched-detection delta 4.3e-04.

Latency on an iPhone 17 Pro running iOS 27.0 beta 6, same graph on both backends, tensor input preallocated, three interleaved blocks of 50 iterations, thermal nominal, block-to-block spread under 1 percent. At fp32, Core AI 5.64 ms against Core ML 6.04 ms. At fp16, Core ML 1.56 ms against Core AI 3.13 ms, though the Core AI fp16 boxes stay within 0.18 px of the fp32 reference where the Core ML fp16 boxes move 0.64 px. On an M4 Max the fp32 ordering reverses: Core ML 2.28 ms, Core AI 2.61 ms.

One caveat on the raw tensors: the 300-row output is order-sensitive, and equal-confidence rows swap between backends. Comparing the raw tensor gives a max delta of 6.3e+02 on a model that is actually correct, so the deltas above come from matching each reference detection to its nearest predicted box instead.

Two things worth knowing if you add this format: an AOT bundle compiled with `--preferred-compute none` will not specialize to CPU-only on device, and Core AI's ANE and GPU preferences land within 6 percent of each other here, which suggests the graph is not fully placed on the ANE yet.

Happy to open a PR adding `format="coreai"` if that shape works for you. I can also run the sweep again against the shipping OS once it lands.

**Comment by john-rocky** [comment]
Correction to my note above: I said the Core AI / Core ML gap suggested the graph was not fully placed on the ANE. That was a guess and it is wrong. I isolated it, and the cause is one op in the end2end head.

Splitting the graph, iPhone 17 Pro on iOS 27.0 beta 6, fp16 on both sides, same input, host-side readback equalized, three interleaved blocks of 50:

| | Core AI | Core ML |
|---|---|---|
| backbone + neck + DFL only (postprocess replaced by identity) | 1.32 ms | 1.33 ms |
| full model | 3.09 ms | 1.55 ms |
| so the end2end postprocess costs | **1.77 ms** | **0.22 ms** |

The body is a tie, within 1 percent. The whole difference is the postprocess.

Inside the postprocess it is the top-k selection. Core AI's op distribution has `sort` and `argsort` and no `topk`, where the Core ML MIL program has `topk` twice. That predicts the cost should be independent of `max_det`, and it is: at `max_det=300` the full model is 3.09 ms, at `max_det=10` it is 3.00 ms. Cutting k by 30x buys 3 percent, so the graph is sorting all 8400 x 80 scores and then slicing, rather than selecting the top k.

One thing that did *not* matter, in case it saves someone the experiment: without `model.fuse()`, the exported ATen graph carries 96 unfused BatchNorms and coreai-torch does not fold them, going from 367 to 602 ops. On device that is worth about 1 percent, so the Core AI compiler evidently folds them itself. The op count is misleading here; only the top-k is real.

So a `format="coreai"` export is fine today for anyone doing `max_det` selection outside the graph, and for the in-graph end2end head it is waiting on a `topk` lowering. I can file that with Apple with this reproducer if it would help.

**Comment by john-rocky** [comment]
Isolated one step further, and the cause is not the sort itself. It is the compute-unit boundary the sort forces.

Same iPhone 17 Pro, iOS 27.0 beta 6, fp16, staged graphs, three interleaved blocks of 40:

| graph | ANE-preferred | GPU-only |
|---|---|---|
| body + max over classes | 1.23 ms | 3.16 ms |
| the same, plus one `topk` | 2.99 ms | 3.20 ms |
| delta from adding `topk` | **+1.75 ms** | **+0.05 ms** |

On the GPU the top-k costs 0.05 ms. It only costs 1.75 ms when the rest of the graph is on the ANE, which is what the earlier evidence was pointing at without me naming it correctly: the cost is invariant to `max_det` (300 vs 10), nearly invariant to axis length (8400 vs 24000 adds 0.06 ms), and replacing `sort`+`argsort` with a single `argsort`+`gather` buys 1.8 percent. All four observations are what a fixed partition-crossing cost looks like, and none of them are what an expensive sort looks like.

Core AI's dialect has `sort`, `argsort`, `argmax`, `gather_along_axis` and no top-k primitive, so `aten.topk` lowers to a full sort of the axis followed by a slice. That lowering is fine on the GPU. What is expensive is that it cannot stay on the ANE, so the graph is cut there.

What this means for a `format="coreai"` export, concretely:

- With the end2end head in the graph: 3.08 ms, against Core ML's 1.53 ms.
- With the head out of the graph (postprocess replaced by identity, top-k done by the caller): **Core AI 1.30 ms, Core ML 1.31 ms.** A tie, and both faster than either full model.

So the body converts and runs at parity today. If you take this format, the export is probably better off emitting the raw `(1, 8400, 84)` predictions and leaving `max_det` selection to the caller, at least until there is a top-k that survives the ANE partition. That also side-steps `aten.remainder.Scalar` entirely, since `(index % nc)` lives in the postprocess.

I am writing this up for Apple as a top-k lowering / partition-boundary report with the ladder above as the reproducer. I will link it here when it is filed. Worth saying plainly: the placement explanation is inferred from the fixed-cost signature and the GPU-only control, not confirmed against a placement API, so Apple may correct the mechanism.

**Comment by john-rocky** [comment]
The PR is up: ultralytics/ultralytics#25926.

It exports all five task heads, round-trips through `YOLO("yolo26n.aimodel")` for inference, and carries the Ultralytics metadata inside the asset so class names, stride and `end2end` survive. 239 lines across 11 files.

The body-vs-head split above is why I left `max_det` selection in the graph rather than out of it: it matches every other format, and moving it is a one-line change if you would rather have the raw predictions. There is one decision in the PR body I deliberately did not make on my own, about whether the `%` fix belongs in the Core AI exporter or in `head.py`.

**Comment by pderrenger** [comment]
Thanks for opening #25926. Please keep the `%` decomposition scoped to the Core AI exporter for now; changing `head.py` affects shared export paths and should only be considered separately with verified cross-backend tests.
